module/validation/seal_validator.go

```diff
@@ -0,0 +1,271 @@
+package validation
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+type sealValidator struct {
+	state            protocol.State
+	assigner         module.ChunkAssigner
+	verifier         module.Verifier
+	seals            storage.Seals
+	headers          storage.Headers
+	payloads         storage.Payloads
+	skipSealValidity bool
+}
+
+func NewSealValidator(state protocol.State, headers storage.Headers, payloads storage.Payloads, seals storage.Seals,
+	assigner module.ChunkAssigner, verifier module.Verifier, skipSealValidity bool) *sealValidator {
+	rv := &sealValidator{
+		state:            state,
+		assigner:         assigner,
+		verifier:         verifier,
+		headers:          headers,
+		seals:            seals,
+		payloads:         payloads,
+		skipSealValidity: skipSealValidity,
+	}
+
+	return rv
+}
+
+func (s *sealValidator) verifySealSignature(aggregatedSignatures *flow.AggregatedSignature,
+	chunk *flow.Chunk, executionResult *flow.ExecutionResult) error {
+	for i, signature := range aggregatedSignatures.VerifierSignatures {
+		signerId := aggregatedSignatures.SignerIDs[i]
+
+		nodeIdentity, err := identityForNode(s.state, chunk.BlockID, signerId)
+
+		if err != nil {
+			return err
+		}
+
+		atst := flow.Attestation{
+			BlockID:           chunk.BlockID,
+			ExecutionResultID: executionResult.ID(),
+			ChunkIndex:        chunk.Index,
+		}
+		atstID := atst.ID()
+		valid, err := s.verifier.Verify(atstID[:], signature, nodeIdentity.StakingPubKey)
+
+		if err != nil {
+			return fmt.Errorf("failed to verify signature: %w", err)
+		}
+
+		if !valid {
+			return engine.NewInvalidInputErrorf("Invalid signature for (%x)", nodeIdentity.NodeID)
+		}
+	}
+
+	return nil
+}
+
+// Validate checks the compliance of the payload seals and returns the last
+// valid seal on the fork up to and including `candidate`. To be valid, we
+// require that seals
+// 1) form a valid chain on top of the last seal as of the parent of `candidate` and
+// 2) correspond to blocks and execution results incorporated on the current fork.
+// 3) has valid signatures for all of its chunks.
+//
+// Note that we don't explicitly check that sealed results satisfy the sub-graph
+// check. Nevertheless, correctness in this regard is guaranteed because:
+//  * We only allow seals that correspond to ExecutionReceipts that were
+//    incorporated in this fork.
+//  * We only include ExecutionReceipts whose results pass the sub-graph check
+//    (as part of ReceiptValidator).
+// => Therefore, only seals whose results pass the sub-graph check will be
+//    allowed.
+func (s *sealValidator) Validate(candidate *flow.Block) (*flow.Seal, error) {
```

Comment:
Is there code change in this function or purely code moving around?

Reply:
it has little changes but mainly it's moved from `sealExtend` of matching engine.